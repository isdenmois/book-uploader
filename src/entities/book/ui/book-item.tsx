import React, { FC, memo } from 'react'
import { StyleSheet } from 'react-native'
import FastImage from 'react-native-fast-image'

import { BookItem as IBookItem } from 'shared/api'
import { getImageUrl } from 'shared/api/book-image'
import { Box, FileIcon, Item, Text } from 'shared/ui'

interface Props {
  item: IBookItem
  onPress: (item: IBookItem) => void
}

export const BookItem: FC<Props> = memo<Props>(
  ({ item, onPress }) => {
    const other = [item.size, item.lang, item.translation].filter(p => p).join(', ')
    const ext = item.ext.replace('.zip', '')
    const imageUrl = getImageUrl(item.type, item.imageUrl)

    return (
      <Item mb={3} onPress={() => onPress(item)} testID='book-item'>
        {!!imageUrl && (
          <Box borderRadius={4} width={40} height={50} overflow='hidden' backgroundColor='searchSelected'>
            <FastImage style={s.image} source={{ uri: imageUrl }} resizeMode='cover' />

            <Text fontSize={12} fontWeight='bold' textAlign='center' color='white' style={s.text}>
              {ext}
            </Text>
          </Box>
        )}
        {!imageUrl && <FileIcon color='searchSelected' text={ext} />}

        <Item.Text suptitle={item.authors} title={item.title} subtitle={other} />
      </Item>
    )
  },
  (prev, next) => prev.item === next.item,
)

const s = StyleSheet.create({
  image: {
    width: 40,
    height: 50,
  },
  text: {
    backgroundColor: '#0005',
    bottom: 0,
    left: 0,
    right: 0,
    position: 'absolute',
  },
})
