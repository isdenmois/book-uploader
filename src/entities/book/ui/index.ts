export * from './book-item'
export * from './book-list'
