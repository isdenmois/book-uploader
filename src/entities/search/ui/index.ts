export * from './search-filters'
export * from './book-list'
