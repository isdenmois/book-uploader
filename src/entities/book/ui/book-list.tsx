import React, { FC } from 'react'
import { FlatList, Image, StyleSheet, View } from 'react-native'

import { BookItem as IBookItem } from 'shared/api'
import { ActivityIndicator, Box, Text } from 'shared/ui'
import { Images } from 'shared/assets'

import { BookItem } from './book-item'

type Props = {
  books: IBookItem[]
  isLoading: boolean
  onDownload: (item: IBookItem) => void
}

export const BookList: FC<Props> = ({ onDownload, books, isLoading }) => {
  if (isLoading) {
    return (
      <Box flex={1} justifyContent='center'>
        <ActivityIndicator color='searchSelected' size='large' />
      </Box>
    )
  }

  if (!books)
    return (
      <View style={s.empty}>
        <Image style={s.image} source={Images.startSearch} fadeDuration={0} resizeMode='contain' />
        <Text style={s.startText} color='search'>
          Start search books to see a result
        </Text>
      </View>
    )

  return (
    <FlatList
      data={books}
      keyExtractor={item => item.link}
      renderItem={({ item }) => <BookItem item={item} onPress={onDownload} />}
      style={s.list}
      contentContainerStyle={s.listContent}
      ListEmptyComponent={
        <View style={s.empty}>
          <Image style={s.image} source={Images.noResult} fadeDuration={0} resizeMode='contain' />
          <Text style={s.emptyText} color='search'>
            Nothing has found
          </Text>
        </View>
      }
    />
  )
}

const s = StyleSheet.create({
  empty: {
    flex: 1,
    marginTop: 30,
  },
  image: {
    width: '100%',
  },
  startText: {
    fontSize: 20,
    textAlign: 'center',
  },
  emptyText: {
    fontSize: 24,
    textAlign: 'center',
    marginTop: 20,
  },
  list: {
    marginTop: 15,
  },
  listContent: {
    paddingHorizontal: 15,
  },
})
