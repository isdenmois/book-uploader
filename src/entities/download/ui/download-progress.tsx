import React from 'react'
import { useStore } from 'effector-react'

import { ProgressBar, Text } from 'shared/ui'
import { $progress } from '../model'

export const DownloadProgress = () => {
  const progress = useStore($progress)
  if (progress < 0) {
    return null
  }

  if (progress === 0) {
    return <Text variant='secondary'>Connection...</Text>
  }

  return <ProgressBar progress={progress} color='searchSelected' />
}
