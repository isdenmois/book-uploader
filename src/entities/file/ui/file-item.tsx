import React, { FC } from 'react'
import { TouchableOpacity } from 'react-native'

import { FileIcon, Item, ProgressBar, TimesIcon } from 'shared/ui'
import { useConfirm } from 'shared/utils'
import type { FileData } from '../model'

interface Props {
  file: FileData
  disabled?: boolean
  onRemove: (file: FileData) => void
  onParse: (file: FileData) => void
  onShare: (file: FileData) => void
}

export const FileItem: FC<Props> = ({ file, disabled, onRemove, onParse, onShare }) => {
  const parseDisabled = disabled || file.isParsed
  const removeFile = useConfirm('Remove file?', file.title || file.id, () => onRemove(file))

  return (
    <Item
      mb={3}
      onPress={() => onShare(file)}
      onLongPress={parseDisabled ? null : () => onParse(file)}
      disabled={disabled}
      progress={!file.error && <ProgressBar progress={file.progress} color='uploadSelected' />}
      testID={'file-item-' + file.id}
    >
      <FileIcon size={34} color='uploadSelected' text={file.ext} />

      <Item.Text suptitle={file.author} title={file.title} error={file.error} />

      {!disabled && (
        <TouchableOpacity onPress={removeFile} testID='removeFile'>
          <TimesIcon />
        </TouchableOpacity>
      )}
    </Item>
  )
}
