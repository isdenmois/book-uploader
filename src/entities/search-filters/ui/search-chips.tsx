import React, { FC } from 'react'
import { ScrollView, StyleSheet, View } from 'react-native'
import { useStore } from 'effector-react'

import { Chip } from 'shared/ui'

import { $searchFilters, setEpub, setFb2, setFlibusta, setZLib } from '../model'

interface Props {
  disabled?: boolean
}

export const SearchChips: FC<Props> = ({ disabled }) => {
  const { extension, type } = useStore($searchFilters)
  const isZLib = type === 'ZLIB'

  return (
    <ScrollView horizontal keyboardShouldPersistTaps='always'>
      <Chip title='Z-Library' selected={type === 'ZLIB'} onPress={setZLib} disabled={disabled} />
      <Chip title='Flibusta' selected={type === 'FLIBUSTA'} onPress={setFlibusta} disabled={disabled} />

      <View style={s.placeholder} />

      {isZLib && <Chip title='EPUB' selected={extension === 'epub'} onPress={setEpub} disabled={disabled} />}
      {isZLib && <Chip title='FB2' selected={extension === 'fb2'} onPress={setFb2} disabled={disabled} />}
    </ScrollView>
  )
}

const s = StyleSheet.create({
  placeholder: { width: 20 },
})
