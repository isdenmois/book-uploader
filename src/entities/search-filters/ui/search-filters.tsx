import React, { FC } from 'react'
import { View } from 'react-native'

import { SearchChips } from './search-chips'
import { SearchQueryInput } from './search-query-input'

interface Props {
  onSubmit()
  disabled?: boolean
}

export const SearchFilters: FC<Props> = ({ onSubmit, disabled }) => {
  return (
    <>
      <SearchQueryInput onSubmit={onSubmit} disabled={disabled} />

      <View>
        <SearchChips disabled={disabled} />
      </View>
    </>
  )
}
