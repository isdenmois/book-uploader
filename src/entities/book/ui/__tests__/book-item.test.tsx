import React from 'react'
import { act, fireEvent, render } from '@testing-library/react-native'

import { BookItem as IBookItem } from 'shared/api'
import { ThemeProvider } from 'shared/ui'

import { BookItem } from '../book-item'

test('BookItem', async () => {
  const item: IBookItem = { ext: 'epub', type: 'ZLIB', title: 'Test result', link: 'http://example.com' }
  const onPress = jest.fn()

  const { toJSON, getByTestId, unmount } = render(<BookItem item={item} onPress={onPress} />, {
    wrapper: ThemeProvider,
  })

  expect(toJSON()).toMatchSnapshot()

  fireEvent.press(getByTestId('book-item'))

  expect(onPress).toHaveBeenCalledWith(item)

  await act(async () => {
    unmount()
  })
})
