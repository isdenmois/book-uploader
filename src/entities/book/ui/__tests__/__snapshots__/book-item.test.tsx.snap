// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookItem 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "display": "flex",
      "marginBottom": 24,
      "opacity": 1,
    }
  }
  testID="book-item"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
      ]
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={40}
      bbWidth={40}
      focusable={false}
      height={40}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "marginLeft": -4,
          },
          Object {
            "flex": 0,
            "height": 40,
            "width": 40,
          },
        ]
      }
      vbHeight={512}
      vbWidth={384}
      width={40}
    >
      <RNSVGGroup>
        <RNSVGGroup>
          <RNSVGPath
            d="M256 0H24A23.94 23.94 0 000 23.88V488a23.94 23.94 0 0023.88 24H360a23.94 23.94 0 0024-23.88V128H272a16 16 0 01-16-16z"
            fill={4283668206}
            propList={
              Array [
                "fill",
              ]
            }
          />
          <RNSVGPath
            d="M384 121.9v6.1H272a16 16 0 01-16-16V0h6.1a24 24 0 0117 7l97.9 98a23.9 23.9 0 017 16.9z"
            fill={4283668206}
            propList={
              Array [
                "fill",
              ]
            }
          />
        </RNSVGGroup>
        <RNSVGText
          content={null}
          dx={Array []}
          dy={Array []}
          fill={4294967295}
          font={
            Object {
              "fontSize": "120",
              "fontWeight": "bold",
              "textAnchor": "middle",
            }
          }
          propList={
            Array [
              "fill",
            ]
          }
          rotate={Array []}
          x={
            Array [
              "192",
            ]
          }
          y={
            Array [
              "440",
            ]
          }
        >
          <RNSVGTSpan
            content="EPUB"
            dx={Array []}
            dy={Array []}
            font={Object {}}
            rotate={Array []}
            x={Array []}
            y={Array []}
          />
        </RNSVGText>
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flex": 1,
            "marginLeft": 8,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#092D5D",
              "display": "flex",
              "fontSize": 16,
            },
          ]
        }
      >
        Test result
      </Text>
    </View>
  </View>
</View>
`;
